### ISSUE TITLE
antlr-ng + official antlr4 java: unexpected predicate error.

### ISSUE THREAD

**Issue by kaby76** [mike-lischke/antlr-ng#30]
Grammar: golang

Runtime: 4.13.1 Antlr Java.

Tool: antlr-ng, commit [f67296c701b7263a303db61677e8aab37312bce5](https://github.com/mike-lischke/antlr-ng/tree/f67296c701b7263a303db61677e8aab37312bce5)

Input: 1_20_version.go

Output:
```
ken@DESKTOP-DL44R7B:~/grammars-v4/golang/Generated-Java$ bash run.sh ../examples/1_20_version.go
line 3:19 rule eos failed predicate: {this.closingBracket()}?
line 9:26 missing '{' at '{'
line 10:9 no viable alternative at input 'A[string]'
line 11:0 extraneous input '}' expecting {<EOF>, 'func', 'const', 'type', 'var'}
line 13:11 mismatched input 'comparable' expecting {'(', '['}
line 14:1 extraneous input 'return' expecting {<EOF>, 'func', 'const', 'type', 'var'}
Java 0 ../examples/1_20_version.go fail 0.113
Total Time: 0.208
ken@DESKTOP-DL44R7B:~/grammars-v4/golang/Generated-Java$ 
```
There should be no parse errors.

I ran the parser driver with `-tokens`, and there are differences in the tokens produced, at least in the text. That shouldn't happen. E.g., Linux: `[@6,24:26='A[T',<27>,3:5]
`; Windows: `[@6,24:24='A',<27>,3:5]`. The serialized ATN table is different between the two generators.

**Comment by mike-lischke** [comment]
That's difficult for me to fix. I don't know Go. If you can reproduce the same problem with JS/TS then there's a chance for me. If parsing fails only for one target then it's like not the code generation, but that can be checked by comparing the output of the code generators. Please check also if the grammar uses Unicode properties (like p{L}). If so this likely fails into the same error class like the other reports.

**Comment by kaby76** [comment]
> That's difficult for me to fix. I don't know Go. If you can reproduce the same problem with JS/TS then there's a chance for me. If parsing fails only for one target then it's like not the code generation, but that can be checked by comparing the output of the code generators. Please check also if the grammar uses Unicode properties (like p{L}). If so this likely fails into the same error class like the other reports.

It's a grammar for Go, but the parser target is Java. [The grammar works for most targets](https://github.com/antlr/grammars-v4/blob/7f917add1199e8b050a7117814899ebab50a3c3b/golang/desc.xml#L4), but I'll have to make a port for Antlr4ng.

**Comment by mike-lischke** [comment]
Porting the GoParserBase to TS is trivial and for me everything parsers fine. I have used the content of that mentioned example:

```go
package onetwenty

type A[T comparable] struct {
	ComparableField T
}

type B = string

type Constraint interface {
	A[string] | ~B
}

func foo[T comparable](value T) A[T] {
	return A[T]{ComparableField: value}
}
```

and got no warning or error.

**Comment by mike-lischke** [comment]
I just see that predication you mentioned `{this.closingBracket()}?` doesn't appear in that example file or any other of the example files. What did you actually use as input?

**Comment by kaby76** [comment]
Here's a JavaScript port of the same code with a driver (Test.js). "build.sh" uses the antlr-ng code generator with JavaScript templates (avoiding the TypeScript templates that use antlr4ng). This program crashes for `bash run.sh ../examples/1_20_version.go.
```
ken@DESKTOP-DL44R7B:~/grammars-v4/golang-js/Generated-JavaScript$ bash run.sh ../examples/1_20_version.go
line 3:19 rule eos failed predicate: {this.closingBracket()}?
line 9:26 missing '{' at '{'
line 10:9 no viable alternative at input 'A[string]'
line 11:0 extraneous input '}' expecting {<EOF>, 'func', 'const', 'type', 'var'}
line 13:11 mismatched input 'comparable' expecting {'(', '['}
line 14:1 extraneous input 'return' expecting {<EOF>, 'func', 'const', 'type', 'var'}
JavaScript 0 ../examples/1_20_version.go fail 0.105
Total Time: 0.105
ken@DESKTOP-DL44R7B:~/grammars-v4/golang-js/Generated-JavaScript$ 
```
The build.sh specifically uses the antlr-ng code generator with the Antlr 4.13.2 JavaScript runtime. 
[golang-js-archive.zip](https://github.com/user-attachments/files/18285860/golang-js-archive.zip)

```
ken@DESKTOP-DL44R7B:~/grammars-v4/golang-js/Generated-JavaScript$ cat build.sh
# Generated from trgen 0.22.0
set -e

if [ -f transformGrammar.py ]; then python3 transformGrammar.py ; fi

# Because there is no integrated build script for Dart targets, we need
# to manually look at the version in package.json and extract the
# version number. We can then use this with antlr4 to generate the
# parser and lexer.
version=`grep antlr4 package.json | awk '{print $2}' | tr -d '"' | tr -d ',' | tr -d '\r' | tr -d '\n'`

tsx --tsconfig ~/antlr-ng/tsconfig.json ~/antlr-ng/cli/runner.ts -Dlanguage=JavaScript   GoLexer.g4
tsx --tsconfig ~/antlr-ng/tsconfig.json ~/antlr-ng/cli/runner.ts -Dlanguage=JavaScript   GoParser.g4


npm install
exit 0
ken@DESKTOP-DL44R7B:~/grammars-v4/golang-js/Generated-JavaScript$
```

The predicate is in two locations.

https://github.com/antlr/grammars-v4/blob/de9bc21fb1134967672dcdd63b68503fd0996684/golang/GoParser.g4#L144 

https://github.com/antlr/grammars-v4/blob/de9bc21fb1134967672dcdd63b68503fd0996684/golang/GoParser.g4#L530

**Comment by mike-lischke** [comment]
I just debugged through the entire parsing process for that input and found no problems, whatsoever. `this.closingBracket()` is called by `adaptivePredict()` and returns false, when parsing the first block in the input. This leads to a prediction of alt 1 in `GoParser.statementList()` (the one for the SEMI symbol, which is however optional, so nothing goes bad there.

The error `line 3:19 rule eos failed predicate: {this.closingBracket()}?` is strange, because there is no rule `eos` (but `EOS`, which however doesn't have that predicate).

I believe the other errors are just follow up issues from the first problem, but since I cannot reproduce that I'm out of ideas here.

**Comment by kaby76** [comment]
I just checked the golang-js-archive.zip I attached, and it's empty. So, you're not debugging what I am generating and running.

**Comment by mike-lischke** [comment]
Correct, I directly used the grammar, generated the parser/lexer and run a script which loads them and parses the mentioned input. Why should I debug anything else? If the unmodified grammar works and your changed version fails, then it's likely the change process, which causes the problem, no?

Do you use a grammar construct that so never appears in the grammar repository (so it wasn't tested so far)?

**Comment by kaby76** [comment]
> Why should I debug anything else? 

Because I don't know what you are debugging??

As I mentioned, parsing works fine with `{ antlr-ng tool -Dlanguage=TypeScript with antlr4ng runtime }`, but there are no templates in antlr-ng for "Official Antlr 4.13.2 TypeScript".

The easiest check would be to do a diff between the generated files from `antlr-ng -Dlanguage=JavaScript` and compare the generated files. This bug feels like a Unicode issue because https://github.com/mike-lischke/antlr-ng/issues/35 is for this grammar, and antlr-ng crashes saying it's a Unicode problem.

**Comment by kaby76** [comment]
Well...I can't get the antlr-ng tool/official JavaScript runtime pair to crash using the latest antlr-ng. And, I *also* cannot get the antlr-ng tool/official Java 4.13.2 to crash! It's likely the recent changes fixed it. So, I'm closing this bug.

But, I still get the `--atn true` of GoLexer.g4 to crash with the invalid unicode issue.
